src/backend_model_instance.cc

```diff
@@ -219,9 +219,7 @@ TritonModelInstance::SetInstances(
           secondary_device.device_id());
     }
     for (int32_t c = 0; c < group.count(); ++c) {
-      std::string instance_name{
-          group.count() > 1 ? group.name() + "_" + std::to_string(c)
-                            : group.name()};
+      std::string instance_name{group.name() + "_" + std::to_string(c)};
```

Comment:
Do we have check that the groups must not share the same name? Otherwise we can have instances with the exact same name?

Reply:
> Do we have check that the groups must not share the same name?

No, we actually allow one or more instances to share the same name. There are two scenarios:
1. One or more instances are the **same** and share the same name.
    - Exactly which instance is reused or removed during an instance update is undefined, because there is no way to tell the difference between those instances.
    - There is a [new test](https://github.com/triton-inference-server/server/pull/6018/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R217-R230) that manipulate instances under this case. The test will pass as long as the number of instances added or removed is correct.
2. One or more instances are **different** but share the same name.
    - The instances are disambiguated by their instance group configuration.
    - There is a [new test](https://github.com/triton-inference-server/server/pull/6018/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R231-R234) that the correct instance is removed and created, if an instance config is changed but still share the same name.

> Otherwise we can have instances with the exact same name?

Yes. They are distinguished by their instance configuration.

Reply:
I think the ability for different instances to have the same name will be confusing in the long term, but is functional now since the logic is based on config and not names. However, with my other comment to add more descriptive  instance names like this:

> What do you think about including device type / kind in the name? ex: "bert_cpu0_instance0", "bert_cpu0_instance1", "bert_gpu0_instance0", etc.

Then I think the names will be unique by default, unless users specify an identical instance group name in two or more models.

Reply:
What is the reason of allowing instances with the same name?

Reply:
It is always allowed. I am not sure on the reason why it is allowed on the first place, in the past.

Reply:
Other than what is currently implemented. I do not think there is a good reason to allow instances with the same name.